File: java/src/main/java/dojo/supermarket/ReceiptPrinter.java
package dojo.supermarket;

import dojo.supermarket.model.*;

import java.util.Locale;

public class ReceiptPrinter {

    private final int columns;

    public ReceiptPrinter() {
        this(40);
    }

    public ReceiptPrinter(int columns) {
        this.columns = columns;
    }

    public String printReceipt(Receipt receipt) {
        StringBuilder result = new StringBuilder();
        for (ReceiptItem item : receipt.getItems()) {
            String price = String.format(Locale.UK, "%.2f", item.getTotalPrice());
            String quantity = presentQuantity(item);
            String name = item.getProduct().getName();
            String unitPrice = String.format(Locale.UK, "%.2f", item.getPrice());

            int whitespaceSize = this.columns - name.length() - price.length();
            String line = name + getWhitespace(whitespaceSize) + price + "\n";

            if (item.getQuantity() != 1) {
                line += "  " + unitPrice + " * " + quantity + "\n";
            }
            result.append(line);
        }
        for (Discount discount : receipt.getDiscounts()) {
            String productPresentation = discount.getProduct().getName();
            String pricePresentation = String.format(Locale.UK, "%.2f", discount.getDiscountAmount());
            String description = discount.getDescription();
            result.append(description);
            result.append("(");
            result.append(productPresentation);
            result.append(")");
            result.append(getWhitespace(this.columns - 3 - productPresentation.length() - description.length() - pricePresentation.length()));
            result.append("-");
            result.append(pricePresentation);
            result.append("\n");
        }
        result.append("\n");
        String pricePresentation = String.format(Locale.UK, "%.2f", (double) receipt.getTotalPrice());
        String total = "Total: ";
        String whitespace = getWhitespace(this.columns - total.length() - pricePresentation.length());
        result.append(total).append(whitespace).append(pricePresentation);
        return result.toString();
    }

    private static String presentQuantity(ReceiptItem item) {
        return ProductUnit.Each.equals(item.getProduct().getUnit())
                ? String.format("%x", (int)item.getQuantity())
                : String.format(Locale.UK, "%.3f", item.getQuantity());
    }

    private static String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }
}

File: dart/supermarket_receipt/lib/supermarket/receipt_printer.dart
import 'package:supermarket_receipt/supermarket/model/product_unit.dart';
import 'package:supermarket_receipt/supermarket/model/receipt.dart';
import 'package:supermarket_receipt/supermarket/model/receipt_item.dart';
import 'package:supermarket_receipt/supermarket/model/discount.dart';

class ReceiptPrinter {
  
  final int _columns;

  ReceiptPrinter() : this._columns = 40;

  ReceiptPrinter.withColumns(this._columns);

  String printReceipt(Receipt receipt) {
    StringBuffer result = StringBuffer();
    for (ReceiptItem item in receipt.getItems()) {
      String receiptItem = presentReceiptItem(item);
      result.write(receiptItem);
    }
    for (Discount discount in receipt.getDiscounts()) {
      String discountPresentation = presentDiscount(discount);
      result.write(discountPresentation);
    }

    result.write('\n');
    result.write(presentTotal(receipt));
    return result.toString();
  }

  String presentReceiptItem(ReceiptItem item) {
    String totalPricePresentation = presentPrice(item.getTotalPrice());
    String name = item.getProduct().getName();

    String line = formatLineWithWhitespace(name, totalPricePresentation);

    if (item.getQuantity() != 1) {
      line += '  ' + presentPrice(item.getPrice()) + ' * ' + presentQuantity(item) + '\n';
    }
    return line;
  }

  String presentDiscount(Discount discount) {
    String name = discount.getDescription() + ' (' + discount.getProduct().getName() + ')';
    String value = presentPrice(discount.getDiscountAmount());

    return formatLineWithWhitespace(name, value);
  }

  String presentTotal(Receipt receipt) {
    String name = 'Total: ';
    String value = presentPrice(receipt.getTotalPrice());
    return formatLineWithWhitespace(name, value);
  }

  String formatLineWithWhitespace(String name, String value) {
    StringBuffer line = StringBuffer();
    line.write(name);
    int whitespaceSize = _columns - name.length - value.length;
    for (int i = 0; i < whitespaceSize; i++) {
      line.write(' ');
    }
    line.write(value);
    line.write('\n');
    return line.toString();
  }

  String presentPrice(double price) {
    return price.toStringAsFixed(2);
  }

  String presentQuantity(ReceiptItem item) {
    if (item.getProduct().getUnit() == ProductUnit.EACH) {
      return item.getQuantity().toInt().toString();
    } else {
      return item.getQuantity().toStringAsFixed(3);
    }
  }
}